### PR TITLE
Make Value Checker aware of the `@Unsigned` annotation

### PR DESCRIPTION
**Pull request by mernst**
It is a bit unfortunate how intertwined this makes the Value Checker and the Signedness Checker.
Should the Signedness Checker be moved into the `common` package?

It would reduce the duplication of methods to always require the boolean `isUnsigned` argument.
That would also reduce the chances of accidentally calling the version without the argument.
Should I make that change?

### CODE REVIEW

File: framework/src/main/java/org/checkerframework/common/value/util/Range.java [hunk 1]
```diff
@@ -344,6 +365,26 @@ public Range intRange() {
...
+   * range of the Short class, return SHORT_EVERYTHING.
+   *
+   * <p>If {@link #ignoreOverflow} is false and the bounds of this range are not representable as
+   * 16-bit integers, convert the bounds to Short type in accordance with Java twos-complement
+   * overflow rules, e.g., Short.MAX_VALUE + 1 is converted to Short.MIN_VALUE.
+   *
+   * @param isUnsigned if true, return a range of unsigned shorts
+   * @return this range, converted to a 16-bit short range
+   */
+  public Range shortRange(boolean isUnsigned) {
```
Comment (smillst) [inline framework/src/main/java/org/checkerframework/common/value/util/Range.java:387]: This `isUnsigned` parameter is not used.

File: framework/src/main/java/org/checkerframework/common/value/util/NumberUtils.java [hunk 2]
```diff
@@ -32,24 +46,81 @@ public static List<? extends Number> castNumbers(
...
 
+  /**
+   * Returns the given number, casted to unsigned byte.
+   *
+   * @param n a number
+   * @return the given number, casted to unsigned byte
+   */
+  private static Short byteValueUnsigned(Number n) {
+    short result = n.byteValue();
+    if (result > Byte.MAX_VALUE) {
```
Comment (smillst) [inline framework/src/main/java/org/checkerframework/common/value/util/NumberUtils.java:90]: `result > Byte.MAX_VALUE` is never true.  (Same comment for the other *ValueUnsigned methods.

File: framework/src/main/java/org/checkerframework/common/value/ValueTreeAnnotator.java [hunk 3]
```diff
@@ -257,6 +258,21 @@ public Void visitTypeCast(TypeCastTree tree, AnnotatedTypeMirror atm) {
...
         return null;
       }
+
+      boolean isUnsigned = false;
+      // I would like to call ((AnnotatedTypeTree) castTree).hasAnnotation(Unsigned.class), but
+      // `Unsigned` is in the checker package and this code is in the common package.
+      List<? extends AnnotationTree> annoTrees =
+          TreeUtils.getExplicitAnnotationTrees(null, tree.getType());
+      List<AnnotationMirror> annos = TreeUtils.annotationsFromTypeAnnotationTrees(annoTrees);
+      for (AnnotationMirror am : annos) {
```
Comment (smillst) [inline framework/src/main/java/org/checkerframework/common/value/ValueTreeAnnotator.java:268]: This for loop can be replaced with:
be `isUnsigned = AnnotationUtils#containsSameByName(annos, "org.checkerframework.checker.signedness.qual.Unsigned")`.

**Review by smillst — CHANGES_REQUESTED** [review]
I a bit confused how the Value Checker is using `@Unsigned`.   There's no explanation about how the Value Checker treats `@Unsigned` differently. 
> It is a bit unfortunate how intertwined this makes the Value Checker and the Signedness Checker.
> Should the Signedness Checker be moved into the common package?

The Value Checker also uses `@NonNegative` in the same way, but the Index Checker isn't in the  

> It would reduce the duplication of methods to always require the boolean isUnsigned argument.
> That would also reduce the chances of accidentally calling the version without the argument.
> Should I make that change?

Right now those parameters aren't being used, so I'm not sure how to answer this question.

File: framework/src/main/java/org/checkerframework/common/value/ValueVisitor.java [hunk 4]
```diff
@@ -324,6 +333,154 @@ public Void visitTypeCast(TypeCastTree tree, Void p) {
...
+          }
+        } else {
+          switch (castTypeKind) {
+            case BYTE:
+              {
+                TreeSet<Byte> castValuesTree =
+                    new TreeSet<Byte>(CollectionsPlume.mapList(Number::byteValue, castValues));
+                TreeSet<Byte> exprValuesTree =
+                    new TreeSet<Byte>(CollectionsPlume.mapList(Number::byteValue, exprValues));
+                return sortedSetContainsAll(castValuesTree, exprValuesTree);
```
Comment (smillst) [inline framework/src/main/java/org/checkerframework/common/value/ValueVisitor.java:382]: These sets are no more than 10 elements long, so it's not really necessary to use this special method.

**Review by smillst — APPROVED** [review]
Just one minor comment that you can take or leave.  Thanks!

File: framework/src/main/java/org/checkerframework/common/value/ValueVisitor.java (hunk 4 above)
Comment (mernst) [reply framework/src/main/java/org/checkerframework/common/value/ValueVisitor.java:382]: For sets of size 4, the method saves over 10% of run time.  That said, I do agree this isn't a bottleneck for the checker.

**Pull request merged**